clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java

```diff
@@ -263,16 +279,59 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
         if (log.isDebugEnabled()) {
             log.debug("final assignment: " + assignment);
         }
-
+        
         return assignment;
     }
 
-    private SortedSet<TopicPartition> getTopicPartitions(Map<String, Integer> partitionsPerTopic) {
-        SortedSet<TopicPartition> allPartitions =
-            new TreeSet<>(Comparator.comparing(TopicPartition::topic).thenComparing(TopicPartition::partition));
-        for (Entry<String, Integer> entry: partitionsPerTopic.entrySet()) {
-            String topic = entry.getKey();
-            for (int i = 0; i < entry.getValue(); ++i) {
+    /**
+     * get the unassigned partition list by computing the difference set of the sortedPartitions(all partitions)
+     * and sortedToBeRemovedPartitions. We use two pointers technique here:
+     *
+     * We loop the sortedPartition, and compare the ith element in sorted toBeRemovedPartitions(i start from 0):
+     *   - if not equal to the ith element, add to unassignedPartitions
+     *   - if equal to the the ith element, get next element from sortedToBeRemovedPartitions
+     *
+     * @param sortedPartitions: sorted all partitions
+     * @param sortedToBeRemovedPartitions: sorted partitions, all are included in the sortedPartitions
+     * @return the partitions don't assign to any current consumers
+     */
+    private List<TopicPartition> getUnassignedPartitions(List<TopicPartition> sortedPartitions,
```

Comment:
adopt the refactor 2 from https://github.com/apache/kafka/pull/10552#discussion_r615228600.

We used to have an SortedSet of unassignedPartitions, with all partitions (ex: 1 million partitions), and loop through current assignment, to remove already assigned partitions, ex: 999,000 of them, so we'll only have 1000 partitions left. However, SortedSet element removing need some time because it needs to find element first, and then, do some tree node movement to maintain sorted. This situation should happen a lot since each rebalance, we should only have small set of changes (ex: 1 consumer dropped), so this is an important improvement.

To refactor it, I used two pointer technique to loop through 2 sorted list: sortedPartitions and sortedToBeRemovedPartitions. And only add the difference set of the 2 lists. The looping and element adding is very fast in ArrayList. So, it improves a lot.

Reply:
Just to be clear, a `removeAll(M partitions)` operation on TreeSet should still be only O(M*logN) since each individual remove is only logN. Even for N = 1 million, logN is under 20. So it scales more with how many partitions are being removed. 

I tried to do the math here and found the time complexity of the original to be (slightly) better on paper, but maybe I missed something or the reality is just different for certain input parameters (bigO time is not an absolute law after all 🙂 ) Maybe you can check my work and then run some tests with and without this specific change (but with all other improvements included).

Let's say the number of consumers is C, the number of partitions assigned to each consumer is M, and the total number of partitions is N

Before:
Loop through all consumers and call (TreeSet) unassignedPartitions.removeAll(assignedPartitions). This is `C * M * logN` (where N will actually decrease down to ~0 by the end of the loop since as you pointed out, most partitions should be reassigned in the _sticky_ assignor)

After:
Loop through all consumers and call (ArrayList) toBeRemovedPartitions.addAll(assignedPartitions). Since addAll has to copy all M elements, this is C * M. After that we call sort(toBeRemovedPartitions), where toBeRemovedPartitions = C * M, so this is C * M * log(C * M). And then finally we call getUnassignedPartitions, which  is O(N). So putting it altogether we have C * M + C * M * log(C*M) + N

It's a little hard to compare these directly but in general we'll have C * M ≈ N (since most partitions are reassigned) in which case the before vs after runtimes can be reduced to O(N<sup>2</sup>log(N)) vs O(2N + N<sup>2</sup>log(N)). The before case is actually better, although they're roughly the same for large N. If this analysis holds up in your experiments then we should just go with whichever one uses less memory and/or has the simplest code. In my (admittedly biased) opinion the original method was the easiest to understand in the code, and objectively speaking it also used less memory since we only need the one `unassignedPartitions` data structure. ie we can just rename `sortedAllPartitions` back to `unassignedPartitions` (or maybe partitionsToBeAssigned is better?) and then we can delete both the `partitionsToBeRemoved` and the second `unassignedPartitions`. Thoughts?

Reply:
Of course we also need to take into account that going back to the TreeSet will mean we lose the improvement in `getTopicPartitions` as well. I did the math there and found it came down to O(NlogN) vs O(2N). And since logN grows so quickly this is like a 10x improvement for 1 million partitions. Adding this all up, the new algorithm does come out _slightly_ on top now:

Before we had O(NlogN + N<sup>2</sup>log(N)) and now we have O(4N + NlogN + N<sup>2</sup>log(N)). Taking logN ≈ 20 means a 5x improvement for the part that scales as N.

  -- which is great, obviously, but the time for `getTopicPartitions` is negligible compared to the time complexity of this loop we discussed above, ie the N<sup>2</sup>term is going to dominate the N term. So we have a tradeoff here between an unknown but possibly small performance improvement that uses a lot more memory, and a somewhat worse algorithm with only the one data structure and slightly cleaner code.

That's a hard tradeoff to comment on without more data. If you could re-run your experiments with all other improvements implemented but without these two things (ie the use of `toBeRemovedPartitions instead of dynamically removing from `unassignedPartitions`, and going back to TreeSet in `getTopicPartitions`) then we can see what the actual performance characteristics are and also get a sense of how much extra memory we're talking. I'd recommend first addressing all my other comments here and re-running with/without this stuff in case those other comments have an impact, so we can be fair to both sides.

Anyways that's my analysis here, but you've been looking at this more carefully, more recently so it's possible I'm missing something here. Interested to hear your take on this

Reply:
> In my (admittedly biased) opinion the original method was the easiest to understand in the code, and objectively speaking it also used less memory

I like your honesty, haha! And yes, I agree wit you.

The timing for `getTopicPartitions` and `remove assigned partitions` is here:
**before:**
```
[2021-04-28 17:39:10,260] ERROR getTopicPartitions took:695 
[2021-04-28 17:39:10,266] ERROR remove assigned took:1 
[2021-04-28 17:39:11,594] ERROR getTopicPartitions took:607
[2021-04-28 17:39:12,889] ERROR remove assigned took:1291
```
**after**
```
[2021-04-28 17:36:02,701] ERROR getTopicPartitions took:41
[2021-04-28 17:36:02,707] ERROR remove assigned took:1 
[2021-04-28 17:36:03,290] ERROR getTopicPartitions took:15
[2021-04-28 17:36:04,204] ERROR remove assigned took:910
```

So, basically, your calculation is correct. The `getTopicPartitions` method has 10x (or even more) faster, and the `removed assigned partitions` takes almost the same time, but **after** is still faster. 

I agree it is bad that **after** needs a lot more memory compared with **before**. The more memory usage is due to the additional `toBeRemovedPartitions` created. I can reduce the memory usage by not creating the `sortedAllPartitions` before the assignor started since I only need it when `getUnassignedPartitions`. In `getUnassignedPartitions`, I was looping 2 partitions (`sortedAllPartitions` & `sortedToBeRemovedPartitions`), now, without `sortedAllPartitions`, I can loop through `sortedAllTopics`, something like this:
```
List<String> allTopics = new ArrayList<>(partitionsPerTopic.keySet());
Collections.sort(allTopics);
for (String topic: allTopics) {
    int partitionCount = partitionsPerTopic.get(topic);
    for (int i = 0; i < partitionCount; i++) {
        if (!(nextPartition.topic().equals(topic) && nextPartition.partition() == i)) {
            ...
         }
     }
}
```
This way, **before** and **after** will have the same memory usage. And the time spent is basically the same:
```
[2021-04-28 17:48:21,694] ERROR remove assigned took:46 
[2021-04-28 17:48:23,405] ERROR remove assigned took:1035
```

Reply:
> the removed assigned partitions takes almost the same time, but after is still faster. (explained below)

I think the reason is:
1. `sort(toBeRemovedPartitions)` is `C * M * log(C * M)` , that's correct in theory! But in our assignment case, usually, the consumer assignment is paritition-sorted (due to the consumer leader create the assignment in order). So, based on the java doc in `List.sort`:

> This implementation is a stable, adaptive, iterative mergesort that requires far fewer than n lg(n) comparisons when the input array is partially sorted, while offering the performance of a traditional mergesort when the input array is randomly ordered.  If the input array is nearly sorted, the implementation requires approximately n comparisons.

2. The nature of `ArrayList` is cache-friendly. That is, when retrieving data, we can get their neighbor data at the same time, it improves the data retrieval from memory. So, this nature makes the ArrayList` iteration and add M elements faster than TreeSet.   

I don't know if there are other potential reason, but the performance is indeed improved by `ArrayList`. 

> an unknown but possibly small performance improvement that uses a lot more memory, and a somewhat worse algorithm with only the one data structure and slightly cleaner code.

I'd say, now, with my previous improvement, the memory usage is basically the same. I'd choose to choose my change to make the assignor faster. After all, faster assignor, faster rebalance.

That's my 2 cents.

Reply:
Also, what we are discussing is not 5% or 10% improvement. It's almost 2x faster in my mac, and in jenkins, it is down from 5 sec to 2 sec , 60% improvement. FYI

Reply:
Thanks for getting some concrete numbers to work with! I suspected the theory would not match the reality due to caching primarily, although I wasn't aware of the improved runtime of sort on a partially-ordered list. That's good to know 😄  And it does make sense in hindsight given the nature of the sorting algorithm. 

I've always found that the reality of array performance with any reasonable caching architecture, compared to theoretically better data structures/algorithms is one of those things that people know and still subconsciously doubt. Probably because most people spent 4 years in college getting theoretical algorithmic runtimes drilled into their heads, and far less time looking into the underlying architecture that powers those algorithms and applies its own optimizations under the hood. It's an interesting psychological observation. There's a great talk on it somewhere but I can't remember the name

Anyways, you just never know until you run the numbers. I'm sure this may vary somewhat with different input parameters but I think I'm convinced, let's stick with this improvement. If someone starts complaining about the memory consumption we can always go back and look for ways to cut down. Thanks for the enlightening discussion

Reply:
Just let me know when you've responded to my other comments and this is ready for another pass. Nice to see such an improvement along with making this the default

Reply:
Thank you very much, @ableegoldman ! I'll address your comments today, thank you :)